test/test_main.py

```diff
@@ -283,19 +293,9 @@ def test_setup_event_links(self):
         """
         return NotImplemented
 
-    addCleanup = unittest.TestCase.addCleanup  # noqa
-    assertEqual = unittest.TestCase.assertEqual  # noqa
-    assertFalse = unittest.TestCase.assertFalse  # noqa
-    assertIn = unittest.TestCase.assertIn  # noqa
-    assertIsNotNone = unittest.TestCase.assertIsNotNone  # noqa
-    assertRaises = unittest.TestCase.assertRaises  # noqa
-    assertRegex = unittest.TestCase.assertRegex  # noqa
-    assertNotIn = unittest.TestCase.assertNotIn  # noqa
-    fail = unittest.TestCase.fail
-    subTest = unittest.TestCase.subTest  # noqa
-
-    def setUp(self):
-        self._setup_charm_dir()
+    @pytest.fixture(autouse=True)
```

Comment:
If we didn't use `autouse` and instead had the relevant tests explicitly request this fixture, how many extra lines would that be? In general, I think making it explicit is much nicer than having to figure out that this fixture ~300 lines deep into the module is changing the behaviour of all the tests.

Reply:
Requesting this feature won't add many lines of code, there are about 13 test cases in `_Testmain`. Since this part of code is in the original `setUp` for all test cases, I suppose we need to request it in all test cases.

There are two ways that I can think of:

1. If we request it as a parameter, it might look weird that a test case has a parameter that is not used and shows up as grey in the IDE. Others might accidentally remove it, too.
2. We can get around it by using `@pytest.mark.usefixtures` I think, but there are multiple test classes extending `_TestMain`, not sure how to make the child classes use a fixture defined in the parent class. Technically, we can move the fixture out of the `_TestMain` so that child classes can request it.

I think in this specific case, maybe autouse is simpler? Or should we go for option 2?

Reply:
TIL `@pytest.mark.usefixtures` :smile: . I think that's the best way to go. Then it's very clear to a reader what's happening. I think it would be nice if we could avoid `autouse` and `conftest` as long as possible.